pennylane/math/single_dispatch.py

```diff
@@ -336,7 +336,15 @@ def _coerce_types_torch(tensors):
     """Coerce a list of tensors to all have the same dtype
     without any loss of information."""
     torch = _i("torch")
-    tensors = [torch.as_tensor(t) for t in tensors]
+
+    # Extract existing set devices, if any
+    device_set = set(t.device for t in tensors if isinstance(t, torch.Tensor))
+    if len(device_set) > 1:
+        raise ValueError("Multiple Torch devices were specified, coercing is not possible.")

```

Comment:
I wonder if it makes sense to match the Torch error message?

```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0!```

Reply:
Why not just stick them all on the GPU if at least one is on the GPU?

Reply:
Frankly, I'd like users to know what they are doing. I think we could make the change, but we'll be putting the burden on us providing convenience for user code that is incorect.

Reply:
It seems Torch is providing different approaches for different operations. Arithmetics (e.g., addition) seems to work okay between a tensor on the GPU and a tensor on the CPU:
```python
import torch

a = torch.tensor(0.3, device='cpu')
b = torch.tensor(0.3, device='cuda')

print(a + b)
```
```
tensor(0.6000, device='cuda:0')
```
But stacking doesn't:
```python
print(torch.stack([a,b]))
```
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument tensors in method wrapper___cat)
```
As we're using `_coerce_types_torch` when calling `qml.math.stack`, I'd rather we kept raising an error for now. Good to keep this behaviour in mind though.

Reply:
Yes I think raising an error here is the way to go.